Assets/Scripts/SlipTime/SlipTimeManager.cs

```diff
@@ -36,57 +36,111 @@ public class SlipTimeManager : MonoBehaviour
         {
-            // Logic for entering SlipTime
-            if (Input.GetButtonDown("Fire2") && !inSlipTime && slipTimeCharges > 0)
+            // Logic for entering SlipTime.
+            if (Input.GetButtonDown("Fire2") && !InSlipTime && slipTimeCharges > 0)
             {
-                inSlipTime = true;
+                InSlipTime = true;
                 slipTimeCoefficient = slipTimeScalar;
                 slipTimeCharges -= 1;
             }
 
-            // Logic for exiting SlipTime at user prompt
-            else if (Input.GetButtonDown("Fire2") && inSlipTime)
+            // Logic for exiting SlipTime at user prompt.
+            else if (Input.GetButtonDown("Fire2") && InSlipTime)
             {
-                inSlipTime = false;
-                slipTimeCounter = 0f;
+                InSlipTime = false;
+                SlipTimeCounter = slipTimeDuration;
                 slipTimeCoefficient = 1f;
+                IsChargingSlipTime = true;
+            }
+
+            // Logic to decrease the active SlipTime duration if the player moves.
+            if (Input.GetButton("Horizontal") || Input.GetButton("Vertical") && InSlipTime)
+            {
+                SlipTimeCounter -= Time.deltaTime * slipTimeMovementPenaltyCoefficient;
             }
 
-            // Logic to keep track of when to exit SlipTime by default
-            if (inSlipTime)
+            // Logic to keep track of when to exit SlipTime by default.
+            if (InSlipTime)
             {
-                slipTimeCounter += Time.deltaTime;
-                if (slipTimeCounter > slipTimeDuration)
+                SlipTimeCounter -= Time.deltaTime;
+                if (SlipTimeCounter <= 0)
                 {
-                    inSlipTime = false;
-                    slipTimeCounter = 0f;
+                    InSlipTime = false;
+                    SlipTimeCounter = slipTimeDuration;
                     slipTimeCoefficient = 1f;
+                    IsChargingSlipTime = true;
                 }
             }
 
-            if (slipTimeCharges < maxSlipTimeCharges && (slipTimeCharges != maxSlipTimeCharges - 1 || !inSlipTime))
+            // Logic to handle the cooldown of SlipTime charges.
+            if (IsChargingSlipTime)
             {
-                chargeTimeCounter += Time.deltaTime;
-                if (chargeTimeCounter >= slipTimeChargeDuration)
+                ChargeTimeCounter -= Time.deltaTime;
+                if (ChargeTimeCounter <= 0)
                 {
                     slipTimeCharges += 1;
-                    chargeTimeCounter = 0f;
+                    ChargeTimeCounter = slipTimeCooldownDuration;
+                    if (slipTimeCharges == maxSlipTimeCharges)
+                    {
+                        IsChargingSlipTime = false;
+                    }
                 }
             }
         }
+
+        /// <summary>
+        /// Decrease the current cooldown of SlipTime by the given amount.
```

Comment:
"given amount" here is a bit confusing since we don't pass any arguments. Can we just say "Decrease the current cooldown of SlipTime by a fixed amount proportional to deltaTime?"

Reply:
Whoops, this is left over from a first pass at the implementation.